File: NSFREVENTR/Q/Q/EnterpriseFileSystem.vb
Public Class EnterpriseFileSystem

    Private _tempFolder As String
    Public ReadOnly Property TempFolder() As String
        Get
            Return _tempFolder
        End Get
    End Property

    Private _ftpFolder As String
    Public ReadOnly Property FtpFolder() As String
        Get
            Return _ftpFolder
        End Get
    End Property

    Private _logsFolder As String
    Public ReadOnly Property LogsFolder() As String
        Get
            Return _logsFolder
        End Get
    End Property

    Private ReadOnly _fileSystemDataAccess As DataAccess

    Public Sub New(ByVal testMode As Boolean, ByVal region As ScriptSessionBase.Region)
        _fileSystemDataAccess = New DataAccess(testMode, region)
        _tempFolder = _fileSystemDataAccess.GetFileSystemPath("TEMP")
        _ftpFolder = _fileSystemDataAccess.GetFileSystemPath("FTP")
        _logsFolder = _fileSystemDataAccess.GetFileSystemPath("LOGS")
    End Sub

    Public Function GetPath(ByVal fileSystemKey As String) As String
        Return _fileSystemDataAccess.GetFileSystemPath(fileSystemKey)
    End Function

End Class

File: C/CommonLibraries/Common/DataAccess/EnterpriseFileSystem.cs
using System;
using System.Collections.Generic;
using System.Data.SqlClient;

namespace Uheaa.Common.DataAccess
{
    public static class EnterpriseFileSystem
    {
        public static string TempFolder
        {
            get
            {
                return GetPath("TEMP");
            }
        }
        public static string FtpFolder
        {
            get
            {
                return GetPath("FTP");
            }
        }
        public static string LogsFolder
        {
            get
            {
                return GetPath("LOGS");
            }
        }

        public static string SessionsFolder
        {
            get
            {
                return GetPath("Sessions");
            }
        }

        private static Dictionary<string, string> FileCache = new Dictionary<string, string>();
        public static string GetPath(string fileSystemKey, DataAccessHelper.Region? region = null)
        {
            var currentRegion = (region ?? DataAccessHelper.CurrentRegion).ToString();
            var cacheKey = currentRegion + ":" + fileSystemKey;
            if (FileCache.ContainsKey(cacheKey)) 
                return FileCache[cacheKey];
            try
            {
                string path = DataAccessHelper.ExecuteSingle<string>("spGENR_GetFileSystemObject", DataAccessHelper.Database.Csys,
                    SqlParams.Single("Key", fileSystemKey), SqlParams.Single("TestMode", DataAccessHelper.TestMode),
                    SqlParams.Single("Region", currentRegion));

                FileCache[cacheKey] = path;
                return path;
            }
            catch (InvalidOperationException ex)
            {
                throw new Exception(string.Format("The file key {0} was not found in EnterpriseFileSystem.  Please contact Systems Support.", fileSystemKey), ex);
            }
        }

        public static void Invalidate()
        {
            FileCache.Clear();
        }
    }
}
